Source/Dafny/Compilers/SinglePassCompiler.cs

```diff
@@ -3164,7 +3164,8 @@ public class CheckHasNoAssumes_Visitor : BottomUpVisitor {
         } else if (DafnyOptions.O.ForbidNondeterminism) {
           Error(s.Tok, "modify statement without a body forbidden by /definiteAssignment:3 option", wr);
         }
-
+      } else if (stmt is HaltRecoveryStatement h) {
+        EmitHaltRecoveryStmt(h.Body, h.HaltMessageVar.CompileName, h.RecoveryBody, wr);
```

Comment:
It IS a bit weird isn't it? All of the `Emit*` methods take individual pieces rather than whole AST nodes and I was just being consistent. I think it's so that multiple callers can emit low-level components with different configuration even if they don't have exactly the corresponding AST node in hand. I'm inclined to leave this as is for consistency unless you feel strongly about it.

Reply:
Okay, I'd prefer the other way but this is fine too.